tests/models/test_grad_norm.py

```diff
@@ -0,0 +1,102 @@
+import torch
+import pytest
+import numpy as np
+
+from pytorch_lightning import Trainer
+
+from pytorch_lightning.loggers import LightningLoggerBase
+from pytorch_lightning.utilities import rank_zero_only
+
+from tests.base import EvalModelTemplate
+
+
+class OnlyMetricsListLogger(LightningLoggerBase):
+    def __init__(self):
+        super().__init__()
+        self.metrics = []
+
+    @rank_zero_only
+    def log_metrics(self, metrics, step):
+        self.metrics.append(metrics)
+
+    @property
+    def experiment(self):
+        return 'test'
+
+    @rank_zero_only
+    def log_hyperparams(self, params):
+        pass
+
+    @rank_zero_only
+    def finalize(self, status):
+        pass
+
+    @property
+    def name(self):
+        return 'name'
+
+    @property
+    def version(self):
+        return '1'
+
+
+class ModelWithManualGradTracker(EvalModelTemplate):
+    def __init__(self, norm_type, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.stored_grad_norms, self.norm_type = [], norm_type
+
+    # validation spoils logger's metrics with `val_loss` records
+    validation_step = None
+    val_dataloader = None
```

Comment:
I see, I think you want the two dicts to be exactly the same, but I think the test is valid if we test that stored_grad_norms is a subset of the one logged. What do you think? This would simplify the test without compromising the validity of it.

Reply:
Not exactly, since I use lists to store norm-dicts.

Logger's `log_metrics` stores whatever the caller requests in a list. I have found relevant two points where `log_metrics` is invoked: in [trainer.run_training_epoch()](https://github.com/PyTorchLightning/pytorch-lightning/blob/0be530a42774a9a7868ab1f7ba3f6392d6fcde9e/pytorch_lightning/trainer/training_loop.py#L483) and in [trainer.run_evaluation()](https://github.com/PyTorchLightning/pytorch-lightning/blob/0be530a42774a9a7868ab1f7ba3f6392d6fcde9e/pytorch_lightning/trainer/evaluation_loop.py#L385). In lockstep with training loop the custom model updates its own list of grad norm dicts in `on_after_backward` called by [trainer.run_training_batch()](https://github.com/PyTorchLightning/pytorch-lightning/blob/0be530a42774a9a7868ab1f7ba3f6392d6fcde9e/pytorch_lightning/trainer/training_loop.py#L609) from [trainer.run_training_epoch()](https://github.com/PyTorchLightning/pytorch-lightning/blob/0be530a42774a9a7868ab1f7ba3f6392d6fcde9e/pytorch_lightning/trainer/training_loop.py#L438)). Hence, in order to keep the logger's list synced with model's list I decided to disable validation. IMO, this is a cleaner solution than manually *deinterleaving* validation dicts and grad norm dicts from logger's list.